keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1190 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This proposes a component leader election mechanism that is safer for upgrades
+and rollbacks.
+
+This leader election approach continues to use leases, but with two key
+modifications:
+
+- Instead of a race by component instances to claim the lease, component
+  instances declare candidacy for a lease and a election coordinator claims the
+  lease for the best available candidate. This allows the election coordinator
+  to pick a candidate with the lowest version to ensure that skew rules are not
+  violated.
+- The election coordinator can mark a lease as "end of term" to signal to the
+  current leader to stop renewing the lease. This allows the election
+  coordinator to preempt the current leader and replace it with a better one.
+
+## Motivation
+
+The most common upgrade approach used for Kubernetes control plane components is
+a node-by-node approach where all the component of a control plane node are
+terminated together and then restarted at the new version. This process is
+performed node-by-node across a high availability configuration. 
+
+Systems using node-by-node upgrades:
+
+- Cluster API
+- kubeadm
+- KIND
+
+To respect the [Kubernetes skew
+policy](https://kubernetes.io/releases/version-skew-policy/):
+
+- Upgrades should keep controller managers and schedulers at the *old* version
+  until all apiservers are upgraded.
+- Rollbacks should rollback controller managers and schedulers at the *old*
+  version before any apiservers are rolledback.
+
+But a node-by-node upgrade or rollback does not achieve this today.
+
+- For 3 node control plane upgrade, there is about a 25% chance of a new version
+  of the controller running while old versions of the apiserver are active,
+  resulting in a skew violation. (Consider case where the 2nd node upgraded has
+  the lease)
+- For rollback, it is almost a certainty that skew will be violated.
+
+There is also the possiblity that the lease will be lost by a leader during an
+upgrade or rollback resulting in the version of the controller flip-flopping
+between old and new.
+
+### Goals
+
+During HA upgrades/rollbacks/downgrades, 
+
+Leader elected components:
+
+- Change versions at predictable times
+- Do not violate version skew, even during node-by-node rollbacks
+
+The control plane:
+
+- Can safely canary components and nodes at the new version for an extended
+  period of time, or to pause an upgrade at any step durning an upgrade. This
+  enhancement, combined with
+  [UVIP](../4020-unknown-version-interoperability-proxy) helps achieve this.
+  
+
+### Non-Goals
+
+- Change the default leader election for components.
+
+## Proposal
+
+- Offer an opt-in leader election mechanism to:
+  - Elect the candidate with the oldest version available.
```

Comment:
why is it a good idea to hardcode this? Makes the mechanism very specific for Kube components with kube's upgrade policy.

Alternatives:
1. let the instances race for being the election coordinator and implement the selection logic in the instances.
2. use some more universal scoring mechanism, like an integer score.

Reply:
I agree this mechanism is very specific for kube components and upgrade policy. This was the intent of the feature though, to deterministic pick a leader while respecting version skew specifically during a rolling upgrade where mixed version components are possible.

Open to discussing alternatives but is there a particular use case you had in mind around something using this that isn't kube component/upgrade policy specific?

Reply:
Many components have similar requirements and suffer the same problems. Think of OpenShift apiservers and controllers, or Crossplane instances. Everything that provides APIs (e.g. by installing CRDs) or webhooks (admission, conversion) should likely be always the newest version (provider), while controllers should be the oldest (consumers).

With that having said, wondering whether providers are usually subject of leader election. In kube the apiservers are not, neither in OpenShift. But potentially there is this proxying happening between them like in kube. Could also be that ordering of complete deployments is actually what one wants, which obviously is out of scope for this KEP.

What I mean with scoring:
```
v1.23.2 => 12302 # newest wins
v1.23.2 => 87697 # oldest wins
```
Of course, one could fake versions too and get the same behaviour:
```
v1.23.2 => v8.76.97
```

tl/dr: we don't lose anything technically speaking by having just kube's logic. If we turned this into a full (non-annotation) API, one could have `preference: Oldest`.

Reply:
That's true. I think we need separate mechanisms for handling each of the cases listed, for instance [UVIP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4020-unknown-version-interoperability-proxy) targets mixed version apiservers with the proxying. Coordinated Leader Election specifically targets leader elected components. Ordering of complete deployments is an interesting idea that probably deserves its own KEP. 

We have explored the scoring idea a bit. IIUC, the client would determine ordering by publishing something like an int in its component lease, and the leader elector can simply pick the largest number. I slightly prefer doing the ordering selection in one centralized place such that if we were to change algorithms, clients are not all expected to update and we can update it server side. If this is done server side, I feel like the conversion vs a comparison logic is quite similar in complexity, but using the versions (1.23.2) instead of the scoring numbers (87697) feels a bit more readable to a human. It could be argued that scoring would be better if we were to calculate this off of multiple fields though and leads to a discussion around extensibility.

re: `preference: Oldest`: IMO this is an extensibility discussion. For the use case we are trying to solve with version skew we are targeting the oldest for all leader elected components. As for `whether providers are usually subject of leader election`, I don't see any examples in kube although that doesn't eliminate the possibility from downstream components from specific providers. @jpbetz have you explored this area? If we were to support the `preferences: XX` option, that also enables the possibility of using addition fields as part of tiebreakers/selection. For instance, looking at the component lease id, etc. I don't mind supporting additional preferences, but I'd like to see a more concrete use case first to ensure supporting extensibility is valuable.

Reply:
I agree that this KEP doesn't solve for the general ordering problems of upgrade orchistration. It's a specific utility that can be useful in specific cases that we see in the ecosystem (kubeadm, ...).

How extensible we make this is something I've only thought about briefly.  I consider `preference: Oldest` to be the most urgently needed option, but I can see utility in preferring newest, preferring a canary instance with some difference in configuration (for 3rd party use cases in particularly), and for preferring a specific controller, purely as a debugging tool ("Somethings wrong, as a cluster admin, I'd like to change KCM leader to X to see what happens").

Reply:
I would definitely like to see more extensibility here - I can definitely imagine cases where we control compatibility story and can guarantee forward compatibility, but we want to limit the number of reelections (e.g. don't have 3 reelections during upgrade, but immediately change to the newest one) because e.g. initialization is very slow or sth like that.
[The current example where this may be the case is cluster-autoscaler].

So yes - can we please add a user-story along those lines and add Beta criteria to reevaluate extensibility?

Reply:
I do expect the extensibility story for this KEP to expand.  Are we OK with making that a beta criteria and focusing on it then?  I feel like having "prefer oldest" as the default functionality for alpha is forward progress and gives us a clear focus for the beta work.